libminifi/src/c2/C2Agent.cpp

```diff
@@ -873,6 +664,161 @@ void C2Agent::update_agent() {
+    }
+  }
+
+  checkTriggers();
+
+  return utils::TaskRescheduleInfo::RetryIn(std::chrono::milliseconds(heart_beat_period_));
+}
+
+utils::TaskRescheduleInfo C2Agent::consume() {
+  const auto consume_success = responses.consume([this] (C2Payload&& payload) {
+    extractPayload(std::move(payload));
+  });
+  if (!consume_success) {
+    extractPayload(C2Payload{ Operation::HEARTBEAT });
+  }
+  return utils::TaskRescheduleInfo::RetryIn(std::chrono::milliseconds(C2RESPONSE_POLL_MS));
+}
+
+utils::optional<std::string> C2Agent::fetchFlow(const std::string& uri) const {
+  if (!utils::StringUtils::startsWith(uri, "http") || protocol_.load() == nullptr) {
+    // try to open the file
+    utils::optional<std::string> content = filesystem_->read(uri);
+    if (content) {
+      return content;
+    }
+  }
+  // couldn't open as file and we have no protocol to request the file from
+  if (protocol_.load() == nullptr) {
+    return {};
+  }
+
+  std::string resolved_url = uri;
+  if (!utils::StringUtils::startsWith(uri, "http")) {
+    std::stringstream adjusted_url;
+    std::string base;
+    if (configuration_->get(minifi::Configure::nifi_c2_flow_base_url, base)) {
+      adjusted_url << base;
+      if (!utils::StringUtils::endsWith(base, "/")) {
+        adjusted_url << "/";
+      }
+      adjusted_url << uri;
+      resolved_url = adjusted_url.str();
+    } else if (configuration_->get("c2.rest.url", base)) {
+      std::string host, protocol;
+      int port = -1;
+      utils::parse_url(&base, &host, &port, &protocol);
+      adjusted_url << protocol << host;
+      if (port > 0) {
+        adjusted_url << ":" << port;
+      }
+      adjusted_url << "/c2/api/" << uri;
+      resolved_url = adjusted_url.str();
+    }
+  }
+
+  C2Payload payload(Operation::TRANSFER, false, true);
+  C2Payload &&response = protocol_.load()->consumePayload(resolved_url, payload, RECEIVE, false);
+
+  auto raw_data = response.getRawData();
+  return std::string(raw_data.data(), raw_data.size());
+}
+
+bool C2Agent::handleConfigurationUpdate(const C2ContentResponse &resp) {
+  auto url = resp.operation_arguments.find("location");
+
+  std::string file_uri;
+  std::string configuration_str;
+
+  if (url == resp.operation_arguments.end()) {
+    logger_->log_debug("Did not have location within %s", resp.ident);
+    auto update_text = resp.operation_arguments.find("configuration_data");
+    if (update_text == resp.operation_arguments.end()) {
+      logger_->log_debug("Neither the config file location nor the data is provided");
+      C2Payload response(Operation::ACKNOWLEDGE, state::UpdateState::SET_ERROR, resp.ident, false, true);
+      response.setRawData("Error while applying flow. Neither the config file location nor the data is provided.");
+      enqueue_c2_response(std::move(response));
+      return false;
+    }
+    configuration_str = update_text->second.to_string();
+  }
+
+  file_uri = url->second.to_string();
```

Comment:
`url` could be `== end` here, if both tests in line 770 and line 773 fail.  I think lines 783-792 should be in an `else` block.

Reply:
nice catch, fixed